pep-9999.txt

```diff
@@ -0,0 +1,171 @@
+directory  per project to have its own independent set of installed packages.
+This will share the standard library with the base installed Python. This will
+avoid the steps to create or activate or deactivate "virtual environments". Any
+Python scripts will use the ``__pypackages__`` from the base directory of the 
+script.
+
+
+
+Motivation
+==========
+
+Python virtual environments have become an essential part of development and
+teaching workflow in the community, but at the same time, they create a
+barrier to entry for many folks. The following are few examples of the issues
+people get into while being introduced to Python (or programming for the first
+time).
+
+- How virtual environments work is a lot of information for anyone new. It takes
+  a lot of extra time and effort to explain them.
+
+- Different platforms and shell environments require different sets of commands
+  to work on the virtual environments. Any workshop or teaching environment with
+  people coming with different operating systems installed on the laptops create a
+  lot of confusion among the participants.
+
+- Virtual environments need to be activated on each opened terminal, if someone
+  creates/opens a new terminal, that by default does not get the same environment
+  as in a previous terminal with virtual environment activated.
+
+
+Specification
+=============
+
+When the Python binary is executed, it attempts to determine its
+prefix (which it stores in ``sys.prefix``), which is then used to find
+the standard library and other key files, and by the ``site`` module
+to determine the location of the site-package directories.  Currently
+the prefix is found (assuming ``PYTHONHOME`` is not set) by first
+walking up the filesystem tree looking for a marker file (``os.py``)
+that signifies the presence of the standard library, and if none is
+found, falling back to the build-time prefix hard coded in the binary.
+
+This PEP proposes to add a new step in this process. If a ``__pypackages__``
+directory is found in the current working directory, then it will be included in
+``sys.path`` after the current working directory and just before the system
+site-packages. This way, if the Python executable starts in the given project
+directory, it will automatically find all the dependencies inside of
+``__pypackages__``.
+
+In case Python scripts, Python will try to find ``__pypackages__`` in the
+base directory of the script. If found, then it will be used, otherwise the
+Python will behave as it does currently.
+
+If ``pip`` finds the ``__pypackages__`` directory in the current directory,
+it will, by default, install any packages there. There will also be a command
+line option to ``pip`` to create the directory in case it is missing.
+
+Projects that use a source management system can include a ``__pypackages__``
+directory (empty or with e.g. a file like ``.gitignore``). After doing a fresh
```

Comment:
I'd remove this paragraph, projects probably shouldn't check in this directory at all and pip should be smart enough to create it (hopefully by default at some point!).

Reply:
Though we may want to document at some point that having the directory there on clone may give you a simpler experience e.g. if you're teaching. But I think we can leave that unsaid here

Reply:
I *think* the ideal final destination is that pip just always defaults to ``--local``, and will create the directory in ``.`` if it can't resolve one somewhere in the hierarchy and require ``--user`` or ``--global`` for other install schemes. 

Of course how to actually get there is an open question and is further complicated by pip supporting older versions of Python for a good long time... but my personal opinion is that's the ideal end state.